access/ms-teams/app.go

```diff
@@ -0,0 +1,517 @@
+
+	a.SpawnCriticalJob(a.mainJob)
+	a.SpawnCriticalJob(a.watcherJob)
+	<-a.Process.Done()
+	return a.Err()
+}
+
+// Err returns the error app finished with.
+func (a *App) Err() error {
+	return trace.Wrap(a.mainJob.Err())
+}
+
+// WaitReady waits for http and watcher service to start up
+func (a *App) WaitReady(ctx context.Context) (bool, error) {
+	return a.mainJob.WaitReady(ctx)
+}
+
+// init initializes the application
+func (a *App) init(ctx context.Context) error {
+	bk := grpcbackoff.DefaultConfig
+	bk.MaxDelay = grpcBackoffMaxDelay
+
+	apiClient, err := client.New(ctx, client.Config{
+		Addrs:       a.conf.Teleport.GetAddrs(),
+		Credentials: a.conf.Teleport.Credentials(),
+		DialOpts: []grpc.DialOption{
+			grpc.WithConnectParams(
+				grpc.ConnectParams{
+					Backoff:           bk,
+					MinConnectTimeout: minConnectTimeout,
+				},
+			),
+		},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.apiClient = apiClient
+	a.pd = pd.NewCAS(
+		apiClient,
+		pluginName,
+		types.KindAccessRequest,
+		EncodePluginData,
+		DecodePluginData,
+	)
+
+	pong, err := a.checkTeleportVersion(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var webProxyAddr string
+	if pong.ServerFeatures.AdvancedAccessWorkflows {
+		webProxyAddr = pong.ProxyPublicAddr
+	}
+
+	a.bot, err = NewBot(a.conf.MSAPI, pong.ClusterName, webProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return a.initBot(ctx)
+}
+
+// initBot initializes bot
+func (a *App) initBot(ctx context.Context) error {
+	log := logger.Get(ctx)
+
+	teamsApp, err := a.bot.GetTeamsApp(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if teamsApp == nil {
+		return trace.NotFound(
+			"MS Teams app not found in org app store. Please, refer to the documentation!",
+		)
+	}
+
+	log.WithField("name", teamsApp.DisplayName).WithField("id", teamsApp.ID).Info("MS Teams app found in org app store")
+
+	if !a.conf.Preload {
+		return nil
+	}
+
+	log.Info("Preloading user data...")
+
+	for _, userID := range a.conf.Recipients.GetAllRecipients() {
+		userData, err := a.bot.FetchUser(ctx, userID)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		log.WithField("user_id", userID).WithField("chat_id", userData.Chat.ID).Info("User found, chat found")
+	}
+
+	log.Info("User data preloaded and cached.")
+
+	return nil
+}
+
+// newWatcherJob creates WatcherJob
+func (a *App) newWatcherJob() lib.ServiceJob {
+	return watcherjob.NewJob(
+		a.apiClient,
+		watcherjob.Config{
+			Watch: types.Watch{
+				Kinds: []types.WatchKind{{Kind: types.KindAccessRequest}},
+			},
+			EventFuncTimeout: handlerTimeout,
+		},
+		a.onWatcherEvent,
+	)
+}
+
+// run starts the main process
+func (a *App) run(ctx context.Context) error {
+	log := logger.Get(ctx)
+
+	ok, err := a.watcherJob.WaitReady(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if ok {
+		log.Info("Plugin is ready")
+	} else {
+		log.Error("Plugin is not ready")
+	}
+
+	a.mainJob.SetReady(ok)
+
+	<-a.watcherJob.Done()
```

Comment:
Should this select on the ctx.Done() also?

Reply:
As far as I remember, it is triggered when the parent context is done. But, just in case, I've added it.